src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1889,6 +2113,15 @@ public sealed class AndSequence : SequenceTests
             {
                 private AndSequence(ImmutableArray<Tests> remainingTests) : base(remainingTests) { }
                 public override Tests Update(ArrayBuilder<Tests> remainingTests) => Create(remainingTests);
+                public static Tests Create(Tests t1, Tests t2)
+                {
```

Comment:
Consider asserting that `t2` is True or an Assignment/Merge.
Ditto for new `OrSequence.Create` below.

Reply:
Assertion added for sideeffect. I don't think it makes sense to do it in this factory method.

Reply:
The reason I was suggesting to add the assertion to the factory method is that when we start using the factory method for new scenarios, we'll want to improve it with more fast cases.

Reply:
(Marked as unresolved to address in the next iteration.)

Reply:
>Consider asserting that t2 is True or an Assignment/Merge.

That's not true, we're also passing `other` which could be any test.

Reply:
Thanks for the clarification. Looking at the call sites again, I think we'll have situations with non-trivial `t1` but True `t2`. Is that correct? If so, let's add a fast-path for that.
Let's take a similar look at the new `OrSequence.Create`.

Reply:
>I think we'll have situations with non-trivial t1 but True t2.

We're either passing `other` or `sideeffect` as t2. I think you're referring to the latter in which case if t1 (precondition) is non-trivial, we definitely have a non-trivial t2 (sideeffect). I can add an assertion to that effect if it helps.